remme/ws/base.py

```diff
@@ -0,0 +1,541 @@
+
+    async def subscribe(self, ws, id, event_type, params):
+        try:
+            evt_tr = EVENT_TRANSLATOR[event_type]
+        except KeyError:
+            raise ClientException(
+                message=f'Event "{event_type}" not defined',
+                msg_id=id)
+
+        subsevt = self._subsevt.setdefault(ws, {})
+        if event_type in subsevt:
+            raise ClientException(
+                message=f'Already subscribed to event "{event_type}"',
+                msg_id=id)
+
+        stream = self._sockets[ws]
+        router = Router(stream)
+
+        event_types = set(subsevt.keys())
+        event_types.add(event_type)
+
+        validated_data = evt_tr.validate(id, params)
+        from_block = validated_data.get('from_block')
+        if not from_block:
+            from_block = (await router.list_blocks(limit=1))['head']
+
+        req_msg = evt_tr.prepare_subscribe_message(event_types, from_block)
+
+        LOGGER.debug(f'Request message: {req_msg}')
+
+        msg = await stream.send(
+            message_type=Message.CLIENT_EVENTS_SUBSCRIBE_REQUEST,
+            message_content=req_msg.SerializeToString(),
+            timeout=ZMQ_CONNECTION_TIMEOUT)
+
+        LOGGER.debug(f'Message type: {msg.message_type}')
+
+        # Validate the response type
+        if msg.message_type != Message.CLIENT_EVENTS_SUBSCRIBE_RESPONSE:
+            raise ClientException(
+                message=f'Unexpected message type {msg.message_type}',
+                msg_id=id)
+
+        # Parse the response
+        response = ClientEventsSubscribeResponse()
+        response.ParseFromString(msg.content)
+
+        # Validate the response status
+        if response.status != ClientEventsSubscribeResponse.OK:
+            if response.status == ClientEventsSubscribeResponse.UNKNOWN_BLOCK:
+                raise ClientException(
+                    message=f'Unknown block "{from_block}"', msg_id=id)
+            raise ClientException(
+                message='Subscription failed: Couldn\'t '
+                        'send multipart', msg_id=id)
+
+        if ws not in self._evthashes:
+            self._evthashes[ws] = set()
+
+        subsevt[event_type] = {
+            'msg_id': id,
+            'validated_data': validated_data,
+        }
+
+        await self.send_msg(ws, 'SUBSCRIBED', id)
+
+    async def unsubscribe(self, ws, id, event_type):
+        subsevt = self._subsevt.get(ws, {})
+        try:
+            del subsevt[event_type]
+        except KeyError:
+            raise ClientException(
+                message='Subscription not found', msg_id=id)
+
+        await self.send_msg(ws, 'UNSUBSCRIBED', id)
+
+    async def consumer(self):
+        while True:
+            LOGGER.info('Consumer: Start fetching a new messages...')
+            datas = []
+            for ws, stream in self._sockets.items():
+                try:
+                    msg = stream.receive_nowait()
```

Comment:
Why? Can't we just wait on the zmq stream?

Reply:
We have different streams, it will be blocked, if somebody has a message and stream what we call does not.